kernel/src/process/pid_namespace/mod.rs

```diff
@@ -0,0 +1,303 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! This module defines the PID namespace.
+//!
+//! PID namespaces are used to isolate threads, processes, process groups and sessions
+//! (we will call these four as `task`s ).
+//! Each task belongs to a unique PID namespace.
+//! Additionally, each PID namespace has its own PID allocator,
+//! which means that tasks in different PID namespaces might have the same PID.
+//!
+//! # Tree Structure
+//!
+//! All PID namespaces form a hierarchical tree structure,
+//! with the init PID namespace as the root.
+//! When the system boots, only a single PID namespace exists,
+//! known as the init PID namespace.
+//! Upon the creation of a new PID namespace,
+//! it becomes a child of the current PID namespace.
+//! This process resembles the creation of processes,
+//! where a newly cloned process becomes a child of the existing process.
+//! Linux imposes a maximum depth limit of 32 for this tree.
+//!
+//! # Isolation
+//!
+//! A task is only visible within its current PID namespace
+//! and all its ancestor namespaces.
+//! Being "visible" means that the process has a unique PID in these namespaces,
+//! allowing it to be the target for system calls such as kill and waitpid,
+//! and to appear in the procfs of these namespaces.
+//!
+
+use core::sync::atomic::{AtomicBool, AtomicU32, AtomicUsize, Ordering};
+
+use ns_pid::{NsPid, TaskId};
+use pid_entry::{PidEntryInner, PID_ENTRY_MAP_LOCK};
+use spin::Once;
+
+use crate::{
+    events::Events,
+    prelude::*,
+    thread::{Thread, Tid},
+};
+
+mod ns_pid;
+mod pid_entry;
+mod process_pid_entries;
+
+pub use ns_pid::AncestorNsPids;
+pub use pid_entry::{PidEntry, PidEntryWithTasklistGuard, TASK_LIST_LOCK};
+pub use process_pid_entries::ProcessPidEntries;
+
+use super::{Pgid, Pid, Process, ProcessGroup, Session, Sid};
+
+/// The PID namespace.
+pub struct PidNamespace {
+    id: usize,
+    nested_level: usize,
+    pid_allocator: AtomicU32,
+    parent: Weak<PidNamespace>,
+    children: Mutex<BTreeMap<usize, Arc<PidNamespace>>>,
+    pid_entry_map: Mutex<BTreeMap<TaskId, Arc<PidEntryInner>>>,
+    is_init_proc_terminated: AtomicBool,
+}
+
+impl PidNamespace {
+    /// Creates a new root namespace.
+    pub fn new_root() -> Self {
+        let id = PID_NS_ID_ALLOCATOR.fetch_add(1, Ordering::Relaxed);
+        Self {
+            id,
+            nested_level: 0,
+            pid_allocator: AtomicU32::new(1),
+            parent: Weak::new(),
+            children: Mutex::new(BTreeMap::new()),
+            pid_entry_map: Mutex::new(BTreeMap::new()),
+            is_init_proc_terminated: AtomicBool::new(false),
+        }
+    }
+
+    /// Creates a new child namespace.
+    pub fn new_child(parent: &Arc<Self>) -> Result<Arc<Self>> {
```

Comment:
Are you sure that all the methods in the `PidNamespace` should be public? I think some of them shouldn't be.